src/app/query/nonce/account-nonces.utils.ts

```diff
@@ -0,0 +1,67 @@
+import { AddressNonces } from '@stacks/blockchain-api-client/lib/generated';
+import { MempoolTransaction, Transaction } from '@stacks/stacks-blockchain-api-types';
+
+export enum NonceTypes {
+  apiSuggestedNonce = 'api-suggested-nonce',
+  clientFallbackNonce = 'client-fallback-nonce',
+  undefinedNonce = 'undefined-nonce',
+}
+
+export function getNextNonce(
+  accountNonces: AddressNonces,
+  confirmedTransactions: Transaction[],
+  pendingTransactions: MempoolTransaction[]
+): { nonce: number; nonceType: NonceTypes } {
+  const lastExecutedNonce = accountNonces.last_executed_tx_nonce;
+  // const lastMempoolTxNonce = accountNonces.last_mempool_tx_nonce;
+  const possibleNextNonce = accountNonces.possible_next_nonce;
+
+  const detectedMissingNonces = accountNonces.detected_missing_nonces;
+  const hasMissingNonces = detectedMissingNonces?.length > 0;
+  const firstMissingNonce = detectedMissingNonces?.sort()[0];
+
+  const hasPendingTxs = pendingTransactions?.length > 0;
+  const pendingNonces = pendingTransactions.map(tx => tx.nonce);
+  const lastPendingTxNonce = pendingTransactions[0]?.nonce;
+  // const lastConfirmedTxNonce = confirmedTransactions[0]?.nonce;
+
+  const pendingNoncesIncludesPossibleNextNonceFromApi = pendingNonces.includes(possibleNextNonce);
+
+  // Validate a 0 nonce returned from the api
+  if (possibleNextNonce === 0 && confirmedTransactions.length > 0)
```

Comment:
Just read your notion document 👍🏼 was helpful

Suggestion on readability: one thing you could do to help a future editor understand this code would be to make descriptive fns for the checks. Even this one for example:

```ts
function apiSuggestsImpossibleZeroNonceAsHasConfirmedTx({ possibleNextNonce, confirmedTransactions });
```

so this becomes
```ts
if (apiSuggestsImpossibleZeroNonceAsHasConfirmedTx({ possibleNextNonce, confirmedTransactions }))
```

Reply:
Good idea, thx, will help make the logic more clear for others. 👍

Reply:
Check out my badass naming updates at some point. 😆 You will be impressed!